### PR TITLE
New Redshift outputter

### PR DESCRIPTION
**Pull request by ChrisPRobinson**
1) Refactor Redshift client to be able to upsert rows
2) Be able to get table schema of specified table

### CODE REVIEW

**Comment by ChrisPRobinson** [comment]
wip

File: lib/etl/redshift/table.rb [hunk 1]
```diff
@@ -47,7 +141,7 @@ def create_table_sql(using_redshift_odbc_driver = true)
         columns.each do |name, column|
           column_type = col_type_str(column)
           column_statement = "\"#{name}\" #{column_type}"
-          column_statement += " IDENTITY(#{@identity_key[:seed]}, #{@identity_key[:step]})" if !@identity_key.empty? && @identity_key[:column] == name.to_sym  
+          column_statement += " IDENTITY(#{@identity_key[:seed]}, #{@identity_key[:step]})" if !@identity_key.empty? && @identity_key[:column] == name.to_sym
           column_statement += " NOT NULL" if @primary_key.include?(name.to_sym) || ( !@identity_key.empty? && @identity_key[:column] == name.to_sym )
```
Comment (emya) [inline lib/etl/redshift/table.rb]: We should use this for column with `nullable` info?

File: lib/etl/transform/row_splitter.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+  # splits the row into multiple named rows
+  class SplitRow < Base
+
+    # column hash map is a hash of arrays of columns.
+    def initialize(column_hash_maps={})
+      super()
+      @column_hash_map = column_hash_map
+      @values = {}
+      values.to_a.each do |v|
```
Comment (emya) [inline lib/etl/transform/row_splitter.rb]: Where does this `values` come from?

File: lib/etl/transform/row_splitter.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        @values[v] = 1
+      end
+    end
+
+    def transform(value)
+      named_rows = {}
+      column_hash_map.each do |row_name, column_names|
```
Comment (emya) [inline lib/etl/transform/row_splitter.rb]: `@column_hash_map`

File: lib/etl/redshift/table.rb (hunk 1 above)
Comment (ChrisPRobinson) [inline lib/etl/redshift/table.rb]: Yes we should use that

File: lib/etl/transform/row_splitter.rb (hunk 2 above)
Comment (ChrisPRobinson) [inline lib/etl/transform/row_splitter.rb]: I fixed this in the latest revision, copy pasta thing

File: lib/etl/transform/row_splitter.rb (hunk 3 above)
Comment (ChrisPRobinson) [inline lib/etl/transform/row_splitter.rb]: fixed

File: lib/etl/redshift/table.rb [hunk 4]
```diff
@@ -31,6 +31,100 @@ def smallint(name, &block)
...
+          character_max = col["character_maximum_length"]
+          width = col["numeric_precision"]
+          scale = col["numeric_scale"]
+          dist_key = col["distkey"]
+          udt_name = col["udt_name"]
+          sort_key = col["sortkey"]
+
+          if pk_info.include?(ordinal_pos)
+            pks << col_name.to_s
+          end
```
Comment (emya) [inline lib/etl/redshift/table.rb]: `pks << col_name.to_s if pk_info.include?` 
is more ruby-like code.

Comment (ChrisPRobinson) [inline lib/etl/redshift/table.rb]: fixed this and other places.

File: spec/redshift/client_spec.rb [hunk 5]
```diff
@@ -34,6 +58,28 @@
...
+        { "id" => 1, "col2" => "value2a" },
+        { "id" => 2, "col2" => "value2b" },
+        { "id" => 3, "col2" => "value2c" },
+      ]
+      input = ETL::Input::Array.new(data)
+      client.upsert_rows(input, ["simple_orgs"])
+      r = client.execute("Select * from simple_orgs")
+      expect(r.ntuples).to eq(3)
+      expect(r.values).to eq([["2", "value2b"], ["3", "value2c"], ["1", "value2a"]])
+    end
```
Comment (emya) [inline spec/redshift/client_spec.rb]: I think we want to have one more test for upserting data with multiple rows that have the same primary key to make sure that upsert and primary keys work.

Comment (ChrisPRobinson) [inline spec/redshift/client_spec.rb]: I added the test, its in there now.

**Pull request merged**